File: native_security/ledger_keyring/android/src/main/kotlin/com/chapter2/ledger/models/Eip712ApprovalPayload.kt
package com.chapter2.ledger.models

public data class Eip712Domain(
    val name: String = "Chapter2",
    val version: String = "1",
    val chainId: Long = 84532L,
    val verifyingContract: String = "0x9b6023D1B6D3b076C8d999Ba406AE486750ce7d3"
) {
    companion object {
        val BASE_SEPOLIA_DEFAULT = Eip712Domain()
    }
}

public data class Eip712ApprovalPayload(
    val actionId: String,
    val agent: String,
    val recipient: String,
    val token: String,
    val amountUnits: ULong,
    val nonce: ULong,
    val deadlineTimestamp: ULong,
    val mandateHash: String,
    val riskScore: UByte,
    val domain: Eip712Domain = Eip712Domain.BASE_SEPOLIA_DEFAULT
) {
    val formattedUsdAmount: String
        get() {
            val dollars = amountUnits.toDouble() / 1_000_000.0
            return String.format("$%.2f", dollars)
        }
}

File: packages/ledger_keyring/ios/Classes/Models/EIP712ApprovalPayload.swift
import Foundation

public struct EIP712Domain: Equatable, Codable, Sendable {
    public let name: String
    public let version: String
    public let chainId: UInt64
    public let verifyingContract: String

    public init(
        name: String = "Chapter2",
        version: String = "1",
        chainId: UInt64 = 84532,
        verifyingContract: String = "0x9b6023D1B6D3b076C8d999Ba406AE486750ce7d3"
    ) {
        self.name = name
        self.version = version
        self.chainId = chainId
        self.verifyingContract = verifyingContract
    }

    public static let baseSepoliaDefault = EIP712Domain()
}

public struct EIP712ApprovalPayload: Equatable, Codable, Sendable {
    public let actionId: String
    public let agent: String
    public let recipient: String
    public let token: String
    public let amountUnits: UInt64
    public let nonce: UInt64
    public let deadlineTimestamp: UInt64
    public let mandateHash: String
    public let riskScore: UInt8
    public let domain: EIP712Domain

    public init(
        actionId: String,
        agent: String,
        recipient: String,
        token: String,
        amountUnits: UInt64,
        nonce: UInt64,
        deadlineTimestamp: UInt64,
        mandateHash: String,
        riskScore: UInt8,
        domain: EIP712Domain = .baseSepoliaDefault
    ) {
        self.actionId = actionId
        self.agent = agent
        self.recipient = recipient
        self.token = token
        self.amountUnits = amountUnits
        self.nonce = nonce
        self.deadlineTimestamp = deadlineTimestamp
        self.mandateHash = mandateHash
        self.riskScore = riskScore
        self.domain = domain
    }

    public var formattedUsdAmount: String {
        let dollars = Double(amountUnits) / 1_000_000.0
        return String(format: "$%.2f", dollars)
    }
}
